In [461]:
import pandas as pd

In [462]:
df=pd.read_csv('/Users/Abdul/Desktop/MMA/Enterprise Data Science/Revenue-Radar/Data/train_dejsonified.csv', parse_dates=['date'])

/var/folders/rj/g3r71fsx2vnbbbjb9v368gr00000gp/T/ipykernel_4465/3898126540.py:1: DtypeWarning: Columns (2,54) have mixed types. Specify dtype option on import or set low_memory=False.
  df=pd.read_csv('/Users/Abdul/Desktop/MMA/Enterprise Data Science/Revenue-Radar/Data/train_dejsonified.csv', parse_dates=['date'])


In [463]:
# Exploring the dataset; checking the data types of the columns
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903653 entries, 0 to 903652
Data columns (total 55 columns):
 #   Column                               Non-Null Count   Dtype         
---  ------                               --------------   -----         
 0   channelGrouping                      903653 non-null  object        
 1   date                                 903653 non-null  datetime64[ns]
 2   fullVisitorId                        903653 non-null  object        
 3   sessionId                            903653 non-null  object        
 4   socialEngagementType                 903653 non-null  object        
 5   visitId                              903653 non-null  int64         
 6   visitNumber                          903653 non-null  int64         
 7   visitStartTime                       903653 non-null  int64         
 8   continent                            903653 non-null  object        
 9   subContinent                         903653 non-null  object        
 

In [464]:
# Checking null values in the dataset
# Note that null values in the transactionRevenue column means that the customer did not make a purchase
df.isnull().sum()

channelGrouping                             0
date                                        0
fullVisitorId                               0
sessionId                                   0
socialEngagementType                        0
visitId                                     0
visitNumber                                 0
visitStartTime                              0
continent                                   0
subContinent                                0
country                                     0
region                                      0
metro                                       0
city                                        0
cityId                                      0
networkDomain                               0
latitude                                    0
longitude                                   0
networkLocation                             0
browser                                     0
browserVersion                              0
browserSize                       

#### Preparing the dataset for the classification model

In [ ]:
df.columns.to_list()

In [466]:
# Checking the number of unique values in each column
df.nunique()

channelGrouping                             8
date                                      366
fullVisitorId                          723803
sessionId                              902755
socialEngagementType                        1
visitId                                886303
visitNumber                               384
visitStartTime                         887159
continent                                   6
subContinent                               23
country                                   222
region                                    376
metro                                      94
city                                      649
cityId                                      1
networkDomain                           28064
latitude                                    1
longitude                                   1
networkLocation                             1
browser                                    54
browserVersion                              1
browserSize                       

In [467]:
# Checking the unique values in the 'bounces' column
df['bounces'].value_counts()

bounces
1.0    450630
Name: count, dtype: int64

In [468]:
# Replacing the null values in the 'bounces' column with 0
df['bounces'].fillna(0, inplace=True)
# Checking the unique values in the 'bounces' column after replacing the null values
df['bounces'].value_counts()

bounces
0.0    453023
1.0    450630
Name: count, dtype: int64

In [469]:
# Finding the columns with constant values
constant_columns = [col for col in df.columns if df[col].nunique() == 1]     
constant_columns

['socialEngagementType',
 'cityId',
 'latitude',
 'longitude',
 'networkLocation',
 'browserVersion',
 'browserSize',
 'operatingSystemVersion',
 'mobileDeviceBranding',
 'mobileDeviceModel',
 'mobileInputSelector',
 'mobileDeviceInfo',
 'mobileDeviceMarketingName',
 'flashVersion',
 'language',
 'screenColors',
 'screenResolution',
 'visits',
 'newVisits',
 'adwordsClickInfo.criteriaParameters',
 'isTrueDirect',
 'adwordsClickInfo.isVideoAd',
 'campaignCode']

### Dealing with Useless Variables

Upon inspecting the dataset to determine the useful columns for the use-case and identifying which columns will be dropped, the following observations were made:

- The following columns are IDs and are not useful for modeling but may be needed for eda. Will drop before modeling
  - `fullVisitorId`
  - `sessionId`
  - `visitId`
  - `cityId`

- The following columns are constant across all observations; therefore, all except `bounces` should be dropped. The `bounces` column contains null values and 1s, indicating that the observations with null values represent users who take actions before exiting:
  - `socialEngagementType`
  - `latitude`
  - `longitude`
  - `networkLocation`
  - `browserVersion`
  - `browserSize`
  - `operatingSystemVersion`
  - `mobileDeviceBranding`
  - `mobileDeviceModel`
  - `mobileInputSelector`
  - `mobileDeviceInfo`
  - `mobileDeviceMarketingName`
  - `flashVersion`
  - `language`
  - `screenColors`
  - `screenResolution`
  - `visits`
  - `newVisits`
  - `adwordsClickInfo.criteriaParameters`
  - `isTrueDirect`
  - `adwordsClickInfo.isVideoAd`
  - `campaignCode`
- The `networkDomain` column has too many unique values; therefore, it should be dropped

In [470]:

# dropping columns with many unique values
df.drop(['networkDomain','visitStartTime','visitNumber'], axis=1, inplace=True)

In [471]:
# Checking the new dimensions of the dataset
df.shape

(903653, 52)

In [472]:

# dropping the columns with constant values
df.drop(constant_columns, axis=1, inplace=True)

In [473]:
# checking the shape of the dataset after dropping the columns with constant values 
df.shape

(903653, 29)

In [474]:
# Calculaing the proportion of missing values in each column
missing_values = df.isnull().mean().sort_values(ascending=False)
missing_values

adContent                         0.987887
transactionRevenue                0.987257
adwordsClickInfo.adNetworkType    0.976252
adwordsClickInfo.slot             0.976252
adwordsClickInfo.page             0.976252
adwordsClickInfo.gclId            0.976140
referralPath                      0.633774
keyword                           0.556551
pageviews                         0.000111
hits                              0.000000
medium                            0.000000
source                            0.000000
campaign                          0.000000
bounces                           0.000000
channelGrouping                   0.000000
date                              0.000000
isMobile                          0.000000
operatingSystem                   0.000000
browser                           0.000000
city                              0.000000
metro                             0.000000
region                            0.000000
country                           0.000000
subContinen

In [475]:
# Checking the number of null values in the dataset
df.isnull().sum()

channelGrouping                        0
date                                   0
fullVisitorId                          0
sessionId                              0
visitId                                0
continent                              0
subContinent                           0
country                                0
region                                 0
metro                                  0
city                                   0
browser                                0
operatingSystem                        0
isMobile                               0
deviceCategory                         0
hits                                   0
pageviews                            100
bounces                                0
transactionRevenue                892138
campaign                               0
source                                 0
medium                                 0
keyword                           502929
referralPath                      572712
adwordsClickInfo

In [476]:
df.head()

channelGrouping       date        fullVisitorId  \
0  Organic Search 2016-09-02  1131660440785968503   
1  Organic Search 2016-09-02   377306020877927890   
2  Organic Search 2016-09-02  3895546263509774583   
3  Organic Search 2016-09-02  4763447161404445595   
4  Organic Search 2016-09-02    27294437909732085   

                        sessionId     visitId continent     subContinent  \
0  1131660440785968503_1472830385  1472830385      Asia     Western Asia   
1   377306020877927890_1472880147  1472880147   Oceania      Australasia   
2  3895546263509774583_1472865386  1472865386    Europe  Southern Europe   
3  4763447161404445595_1472881213  1472881213      Asia   Southeast Asia   
4    27294437909732085_1472822600  1472822600    Europe  Northern Europe   

          country                         region  \
0          Turkey                          Izmir   
1       Australia  not available in demo dataset   
2           Spain            Community of Madrid   
3       Indonesia  not available in demo dataset   
4  United Kingdom  not available in demo dataset   

                           metro  ...   campaign  source   medium  \
0                      (not set)  ...  (not set)  google  organic   
1  not available in demo dataset  ...  (not set)  google  organic   
2                      (not set)  ...  (not set)  google  organic   
3  not available in demo dataset  ...  (not set)  google  organic   
4  not available in demo dataset  ...  (not set)  google  organic   

           keyword referralPath  adwordsClickInfo.page  adwordsClickInfo.slot  \
0   (not provided)          NaN                    NaN                    NaN   
1   (not provided)          NaN                    NaN                    NaN   
2   (not provided)          NaN                    NaN                    NaN   
3  google + online          NaN                    NaN                    NaN   
4   (not provided)          NaN                    NaN                    NaN   

   adwordsClickInfo.gclId  adwordsClickInfo.adNetworkType adContent  
0                     NaN                             NaN       NaN  
1                     NaN                             NaN       NaN  
2                     NaN                             NaN       NaN  
3                     NaN                             NaN       NaN  
4                     NaN                             NaN       NaN  

[5 rows x 29 columns]

In [477]:
df.pageviews.fillna(0, inplace=True)

In [478]:
df.isna().sum()

channelGrouping                        0
date                                   0
fullVisitorId                          0
sessionId                              0
visitId                                0
continent                              0
subContinent                           0
country                                0
region                                 0
metro                                  0
city                                   0
browser                                0
operatingSystem                        0
isMobile                               0
deviceCategory                         0
hits                                   0
pageviews                              0
bounces                                0
transactionRevenue                892138
campaign                               0
source                                 0
medium                                 0
keyword                           502929
referralPath                      572712
adwordsClickInfo

In [479]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 903653 entries, 0 to 903652
Data columns (total 29 columns):
 #   Column                          Non-Null Count   Dtype         
---  ------                          --------------   -----         
 0   channelGrouping                 903653 non-null  object        
 1   date                            903653 non-null  datetime64[ns]
 2   fullVisitorId                   903653 non-null  object        
 3   sessionId                       903653 non-null  object        
 4   visitId                         903653 non-null  int64         
 5   continent                       903653 non-null  object        
 6   subContinent                    903653 non-null  object        
 7   country                         903653 non-null  object        
 8   region                          903653 non-null  object        
 9   metro                           903653 non-null  object        
 10  city                            903653 non-null  object 

In [480]:
# checking the number of unique visitor ids
df.fullVisitorId.nunique()

723803

In [482]:
df['transactionRevenue'].fillna(0, inplace=True)

In [483]:
df['Conversion'] = df['transactionRevenue'].apply(lambda x: 1 if x > 0 else 0)

In [ ]:
# checking the number of conversions for each user
coversion_by_user=dict(df.groupby('fullVisitorId')['Conversion'].sum().sort_values(ascending=False))
coversion_by_user

In [485]:
df['at_least_one_conversion']=df.groupby('fullVisitorId')['Conversion'].transform('sum').apply(lambda x: 1 if x>0 else 0)

In [486]:
df.channelGrouping.value_counts()

channelGrouping
Organic Search    381561
Social            226117
Direct            143026
Referral          104838
Paid Search        25326
Affiliates         16403
Display             6262
(Other)              120
Name: count, dtype: int64

In [ ]:
# checking the number of conversions for each user
conversions=pd.DataFrame(coversion_by_user.items(), columns=['fullVisitorId', 'Conversion'])
conversions

In [488]:
# Ensuring the DataFrame is sorted by fullVisitorId and date
df.sort_values(by=['fullVisitorId', 'date'], inplace=True)

In [489]:
# Grouping by fullVisitorId and get the first and last visit's channelGrouping
first_visit_channel = df.groupby('fullVisitorId')['channelGrouping'].first().reset_index(name='FirstChannelVisit')
last_visit_channel = df.groupby('fullVisitorId')['channelGrouping'].last().reset_index(name='LastChannelVisit')

In [490]:
# Merge the first and last visit channel dataframes
channels_first_last = pd.merge(first_visit_channel, last_visit_channel, on='fullVisitorId', how='inner')
channels_first_last.sample(40) ## other columns will be merged to this dataframe
channels_first_last.shape

(723803, 3)

In [491]:
df_unique = df.drop_duplicates(subset='fullVisitorId', keep='first')
df_unique.shape

(723803, 31)

In [492]:
channels_first_last = pd.merge(channels_first_last, df_unique[['fullVisitorId', 'at_least_one_conversion', 'country','continent','subContinent']].drop_duplicates(), on='fullVisitorId', how='inner')
channels_first_last.shape

(723803, 7)

In [493]:
# checking which users had different first and last channel visits
channels_first_last[channels_first_last['FirstChannelVisit'] != channels_first_last['LastChannelVisit']]

fullVisitorId FirstChannelVisit LastChannelVisit  \
120         357767462604476            Direct          Display   
121         358027242270102       Paid Search   Organic Search   
184         562678147042735    Organic Search         Referral   
219         650183049109374    Organic Search      Paid Search   
229         677695778949032            Direct         Referral   
...                     ...               ...              ...   
723415  9913166595926807960    Organic Search         Referral   
723519  9937314053887872916        Affiliates   Organic Search   
723581  9952660341541977540    Organic Search         Referral   
723620  9960561123114745936    Organic Search         Referral   
723648  9965616755471685787    Organic Search           Social   

        at_least_one_conversion        country continent      subContinent  
120                           0  United States  Americas  Northern America  
121                           0  United States  Americas  Northern America  
184                           1  United States  Americas  Northern America  
219                           0          India      Asia     Southern Asia  
229                           0  United States  Americas  Northern America  
...                         ...            ...       ...               ...  
723415                        0  United States  Americas  Northern America  
723519                        0         Canada  Americas  Northern America  
723581                        0  United States  Americas  Northern America  
723620                        0          Spain    Europe   Southern Europe  
723648                        0      Hong Kong      Asia      Eastern Asia  

[17296 rows x 7 columns]

### Most users had the same first and last channel visits except for 17296 users who had different first and last channel visits

In [ ]:
# creating another variable for total number of visits for each user
total_visits_by_user=dict(df['fullVisitorId'].value_counts().reset_index(name='TotalVisits').values)
total_visits_by_user

In [ ]:
# Creating a variable to capture number of visits through each channel for each user
channel_visits_by_user=dict(df.groupby('fullVisitorId')['channelGrouping'].value_counts())
channel_visits_by_user

In [ ]:
# converting the dictionary to a list of dictionaries
list_of_dicts = []
for (visitor_id, channel), visits in channel_visits_by_user.items():
    list_of_dicts.append({'fullVisitorId': visitor_id, channel: visits})

# Creating a DataFrame
df_channel_visits = pd.DataFrame(list_of_dicts)

# Combining all dictionaries representing the same fullVisitorId
df_channel_visits = df_channel_visits.groupby('fullVisitorId', as_index=False).sum()

df_channel_visits

In [ ]:
# creating another variable for total number of pageviews for each user 
total_pageviews_by_user=dict(df.groupby('fullVisitorId')['pageviews'].sum().reset_index(name='TotalPageviews').values)
total_pageviews_by_user

In [ ]:
# creating a variable for total number of bounces for each user
total_bounces_by_user=dict(df.groupby('fullVisitorId')['bounces'].sum().reset_index(name='TotalBounces').values)
total_bounces_by_user

In [ ]:
# Creating another variable for number of visters per device category
visits_by_device=dict(df.groupby('fullVisitorId')['deviceCategory'].value_counts())
visits_by_device

In [ ]:
# converting the dictionary to a list of dictionaries
list_of_dicts = []
for (visitor_id, device), visits in visits_by_device.items():
    list_of_dicts.append({'fullVisitorId': visitor_id, device: visits})

# Creating a DataFrame
df_visits_by_device = pd.DataFrame(list_of_dicts)

# Combining all dictionaries representing the same fullVisitorId
df_visits_by_device = df_visits_by_device.groupby('fullVisitorId', as_index=False).sum()

df_visits_by_device

In [501]:
# Creating another variable for the first and last session pageviews for each user   
first_session_pageviews=dict(df.groupby('fullVisitorId')['pageviews'].first().reset_index(name='FirstSessionPageviews').values)
last_session_pageviews=dict(df.groupby('fullVisitorId')['pageviews'].last().reset_index(name='LastSessionPageviews').values)

In [502]:
df['campaign'].value_counts()   

campaign
(not set)                                          865347
Data Share Promo                                    16403
AW - Dynamic Search Ads Whole Site                  14244
AW - Accessories                                     7070
test-liyuhz                                           392
AW - Electronics                                       96
Retail (DO NOT EDIT owners nophakun and tianyu)        50
AW - Apparel                                           46
All Products                                            4
Data Share                                              1
Name: count, dtype: int64

In [503]:
df['campaign']=df['campaign'].apply(lambda x: 0 if x=='(not set)' else 1)

In [504]:
df['campaign'].value_counts()   

campaign
0    865347
1     38306
Name: count, dtype: int64

In [ ]:
# creating a variable for number of campaign visits for each user
campaign_visits_by_user=dict(df.groupby('fullVisitorId')['campaign'].sum().reset_index(name='CampaignVisits').values)
campaign_visits_by_user

In [521]:

# Converting dictionaries to DataFrames
df_total_bounces = pd.DataFrame(list(total_bounces_by_user.items()), columns=['fullVisitorId', 'TotalBounces'])
df_total_visits = pd.DataFrame(list(total_visits_by_user.items()), columns=['fullVisitorId', 'TotalVisits'])
df_total_pageviews = pd.DataFrame(list(total_pageviews_by_user.items()), columns=['fullVisitorId', 'TotalPageviews'])
df_first_session_pageviews = pd.DataFrame(list(first_session_pageviews.items()), columns=['fullVisitorId', 'FirstSessionPageviews'])
df_last_session_pageviews = pd.DataFrame(list(last_session_pageviews.items()), columns=['fullVisitorId', 'LastSessionPageviews'])
df_campaign_visits = pd.DataFrame(list(campaign_visits_by_user.items()), columns=['fullVisitorId', 'CampaignVisits'])

# Start with the channels_first_last DataFrame
final_df = channels_first_last

#merging each DataFrame on 'fullVisitorId'
for dfr in [df_total_visits, df_total_bounces, df_channel_visits, df_total_pageviews, df_visits_by_device, df_first_session_pageviews, df_last_session_pageviews, conversions, df_campaign_visits]:
    final_df = pd.merge(final_df, dfr, on='fullVisitorId', how='inner')


final_df.head()

fullVisitorId FirstChannelVisit LastChannelVisit  at_least_one_conversion  \
0   4823595352351    Organic Search   Organic Search                        0   
1   5103959234087    Organic Search   Organic Search                        0   
2  10278554503158    Organic Search   Organic Search                        0   
3  20424342248747    Organic Search   Organic Search                        0   
4  26722803385797    Organic Search   Organic Search                        0   

          country continent      subContinent  TotalVisits  TotalBounces  \
0           China      Asia      Eastern Asia            1           1.0   
1   United States  Americas  Northern America            1           0.0   
2     New Zealand   Oceania       Australasia            1           0.0   
3            Peru  Americas     South America            1           0.0   
4  United Kingdom    Europe   Northern Europe            1           0.0   

   Organic Search  ...  Display  (Other)  TotalPageviews  desktop  mobile  \
0             1.0  ...      0.0      0.0             1.0      1.0     0.0   
1             1.0  ...      0.0      0.0             8.0      0.0     1.0   
2             1.0  ...      0.0      0.0             8.0      1.0     0.0   
3             1.0  ...      0.0      0.0            13.0      1.0     0.0   
4             1.0  ...      0.0      0.0             2.0      1.0     0.0   

   tablet  FirstSessionPageviews  LastSessionPageviews  Conversion  \
0     0.0                    1.0                   1.0           0   
1     0.0                    8.0                   8.0           0   
2     0.0                    8.0                   8.0           0   
3     0.0                   13.0                  13.0           0   
4     0.0                    2.0                   2.0           0   

   CampaignVisits  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 25 columns]

In [524]:
final_df.shape

(723803, 25)

In [523]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 723803 entries, 0 to 723802
Data columns (total 25 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   fullVisitorId            723803 non-null  object 
 1   FirstChannelVisit        723803 non-null  object 
 2   LastChannelVisit         723803 non-null  object 
 3   at_least_one_conversion  723803 non-null  int64  
 4   country                  723803 non-null  object 
 5   continent                723803 non-null  object 
 6   subContinent             723803 non-null  object 
 7   TotalVisits              723803 non-null  int64  
 8   TotalBounces             723803 non-null  float64
 9   Organic Search           723803 non-null  float64
 10  Social                   723803 non-null  float64
 11  Direct                   723803 non-null  float64
 12  Paid Search              723803 non-null  float64
 13  Referral                 723803 non-null  float64
 14  Affi

In [525]:
for col in final_df.select_dtypes(include='float').columns:
    final_df[col] = final_df[col].astype('int')
final_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 723803 entries, 0 to 723802
Data columns (total 25 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   fullVisitorId            723803 non-null  object
 1   FirstChannelVisit        723803 non-null  object
 2   LastChannelVisit         723803 non-null  object
 3   at_least_one_conversion  723803 non-null  int64 
 4   country                  723803 non-null  object
 5   continent                723803 non-null  object
 6   subContinent             723803 non-null  object
 7   TotalVisits              723803 non-null  int64 
 8   TotalBounces             723803 non-null  int64 
 9   Organic Search           723803 non-null  int64 
 10  Social                   723803 non-null  int64 
 11  Direct                   723803 non-null  int64 
 12  Paid Search              723803 non-null  int64 
 13  Referral                 723803 non-null  int64 
 14  Affiliates          

In [526]:
final_df.head()

fullVisitorId FirstChannelVisit LastChannelVisit  at_least_one_conversion  \
0   4823595352351    Organic Search   Organic Search                        0   
1   5103959234087    Organic Search   Organic Search                        0   
2  10278554503158    Organic Search   Organic Search                        0   
3  20424342248747    Organic Search   Organic Search                        0   
4  26722803385797    Organic Search   Organic Search                        0   

          country continent      subContinent  TotalVisits  TotalBounces  \
0           China      Asia      Eastern Asia            1             1   
1   United States  Americas  Northern America            1             0   
2     New Zealand   Oceania       Australasia            1             0   
3            Peru  Americas     South America            1             0   
4  United Kingdom    Europe   Northern Europe            1             0   

   Organic Search  ...  Display  (Other)  TotalPageviews  desktop  mobile  \
0               1  ...        0        0               1        1       0   
1               1  ...        0        0               8        0       1   
2               1  ...        0        0               8        1       0   
3               1  ...        0        0              13        1       0   
4               1  ...        0        0               2        1       0   

   tablet  FirstSessionPageviews  LastSessionPageviews  Conversion  \
0       0                      1                     1           0   
1       0                      8                     8           0   
2       0                      8                     8           0   
3       0                     13                    13           0   
4       0                      2                     2           0   

   CampaignVisits  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 25 columns]

In [527]:
#saving the final DataFrame to a csv file
final_df.to_csv('/Users/Abdul/Desktop/MMA/Enterprise Data Science/final_df.csv', index=False)

In [518]:
final_df.at_least_one_conversion.value_counts()

at_least_one_conversion
0    713689
1     10114
Name: count, dtype: int64

In [520]:
final_df.columns.to_list()

['fullVisitorId',
 'FirstChannelVisit',
 'LastChannelVisit',
 'at_least_one_conversion',
 'country',
 'continent',
 'subContinent',
 'TotalVisits',
 'Organic Search',
 'Social',
 'Direct',
 'Paid Search',
 'Referral',
 'Affiliates',
 'Display',
 '(Other)',
 'TotalPageviews',
 'desktop',
 'mobile',
 'tablet',
 'FirstSessionPageviews',
 'LastSessionPageviews',
 'Conversion',
 'CampaignVisits']